File: ezhil-android/app/src/test/java/com/ezhil/app/ml/ScreeningHeuristicTest.kt
package com.ezhil.app.ml

import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import kotlin.math.abs

/**
 * The screening score must respond to the recording.
 *
 * A previous version returned a flat 0.5 for every input, so every child was
 * scored identically and every child was tagged "phoneme_confusion". A
 * constant is worse than no score at all: it looks like a measurement while
 * carrying no information, and a teacher acts on it.
 */
class ScreeningHeuristicTest {

    /** Mean-square for a sine of the given peak amplitude. */
    private fun meanSquare(amplitude: Float) = amplitude * amplitude / 2f

    @Test
    fun `a clear reader scores well below the phoneme-confusion threshold`() {
        // 0.40 is the tag threshold in ScreeningModel.buildResult.
        val clear = ScreeningHeuristic.score(meanSquare(0.3f), longPauses = 0)
        assertTrue(
            "clear reading scored ${clear.phonemeErrorRate}, which would tag every fluent child",
            clear.phonemeErrorRate < 0.40f,
        )
    }

    @Test
    fun `quieter audio scores higher than clear audio`() {
        val clear = ScreeningHeuristic.score(meanSquare(0.3f), 0)
        val quiet = ScreeningHeuristic.score(meanSquare(0.02f), 0)
        assertTrue(
            "quiet=${quiet.risk} should exceed clear=${clear.risk}",
            quiet.risk > clear.risk,
        )
    }

    @Test
    fun `the score is not a constant across realistic recordings`() {
        val amplitudes = listOf(0.9f, 0.5f, 0.3f, 0.1f, 0.03f, 0.008f)
        val risks = amplitudes.map { ScreeningHeuristic.score(meanSquare(it), 0).risk }
        val spread = (risks.max() - risks.min())
        assertTrue(
            "risk spread was $spread across $amplitudes — the score carries no information",
            spread > 0.1f,
        )
    }

    @Test
    fun `pauses raise the syllable-skip component`() {
        val none = ScreeningHeuristic.score(meanSquare(0.3f), longPauses = 0)
        val many = ScreeningHeuristic.score(meanSquare(0.3f), longPauses = 8)
        assertTrue(many.syllableSkipRate > none.syllableSkipRate)
        assertTrue(many.risk > none.risk)
    }

    @Test
    fun `hesitation outweighs volume in the risk score`() {
        // A loud but halting reader must not score below a quiet but fluent
        // one. An even split allowed exactly that.
        val loudHalting = ScreeningHeuristic.score(meanSquare(0.5f), longPauses = 6)
        val quietFluent = ScreeningHeuristic.score(meanSquare(0.03f), longPauses = 0)
        assertTrue(
            "loud-halting=${loudHalting.risk} should exceed quiet-fluent=${quietFluent.risk}",
            loudHalting.risk > quietFluent.risk,
        )
    }

    @Test
    fun `quietness is bounded and oriented correctly`() {
        assertEquals(0f, ScreeningHeuristic.quietness(meanSquare(1.0f)), 0.001f)
        assertEquals(1f, ScreeningHeuristic.quietness(0f), 0.001f)
        assertTrue(ScreeningHeuristic.quietness(meanSquare(0.02f)) > ScreeningHeuristic.quietness(meanSquare(0.3f)))
    }

    @Test
    fun `every component stays inside its documented range`() {
        for (a in listOf(0f, 0.001f, 0.05f, 0.3f, 1.0f)) {
            for (p in listOf(0, 1, 5, 40)) {
                val s = ScreeningHeuristic.score(meanSquare(a), p)
                assertTrue(s.risk in 0f..1f)
                assertTrue(s.phonemeErrorRate in 0f..0.5f)
                assertTrue(s.syllableSkipRate in 0f..0.5f)
                // Risk is weighted toward hesitation, not volume.
                val expected = 0.75f * s.syllableSkipRate + 0.25f * s.phonemeErrorRate
                assertTrue(abs(s.risk - expected) < 1e-5f)
            }
        }
    }
}

File: ezhil-web/src/lib/__tests__/screeningHeuristic.test.ts
import { describe, expect, it } from 'vitest';

import { errorTags, quietness, riskLevel, score } from '../screeningHeuristic';

/**
 * The screening score must respond to the recording.
 *
 * A previous version returned a flat 0.5 for every input, so every child was
 * scored identically and every child was tagged "phoneme_confusion". A
 * constant is worse than no score at all: it looks like a measurement while
 * carrying no information, and a teacher acts on it.
 *
 * These mirror ScreeningHeuristicTest.kt. If one side is changed alone, the
 * web and the handset will disagree about the same child.
 */

/** Mean-square for a sine of the given peak amplitude. */
const meanSquare = (amplitude: number) => (amplitude * amplitude) / 2;

describe('screening heuristic', () => {
  it('scores a clear reader well below the phoneme-confusion threshold', () => {
    // 0.40 is the tag threshold in errorTags().
    expect(score(meanSquare(0.3), 0).phonemeErrorRate).toBeLessThan(0.4);
  });

  it('scores quieter audio higher than clear audio', () => {
    expect(score(meanSquare(0.02), 0).risk).toBeGreaterThan(score(meanSquare(0.3), 0).risk);
  });

  it('is not a constant across realistic recordings', () => {
    const risks = [0.9, 0.5, 0.3, 0.1, 0.03, 0.008]
      .map(a => score(meanSquare(a), 0).risk);
    expect(Math.max(...risks) - Math.min(...risks)).toBeGreaterThan(0.1);
  });

  it('raises the syllable-skip component with pauses', () => {
    const none = score(meanSquare(0.3), 0);
    const many = score(meanSquare(0.3), 8);
    expect(many.syllableSkipRate).toBeGreaterThan(none.syllableSkipRate);
    expect(many.risk).toBeGreaterThan(none.risk);
  });

  it('weighs hesitation above volume', () => {
    // A loud but halting reader must not score below a quiet but fluent one.
    // An even split allowed exactly that.
    const loudHalting = score(meanSquare(0.5), 6);
    const quietFluent = score(meanSquare(0.03), 0);
    expect(loudHalting.risk).toBeGreaterThan(quietFluent.risk);
  });

  it('keeps quietness bounded and oriented correctly', () => {
    expect(quietness(meanSquare(1.0))).toBeCloseTo(0, 3);
    expect(quietness(0)).toBeCloseTo(1, 3);
    expect(quietness(meanSquare(0.02))).toBeGreaterThan(quietness(meanSquare(0.3)));
  });

  it('keeps every component inside its documented range', () => {
    for (const a of [0, 0.001, 0.05, 0.3, 1.0]) {
      for (const p of [0, 1, 5, 40]) {
        const s = score(meanSquare(a), p);
        expect(s.risk).toBeGreaterThanOrEqual(0);
        expect(s.risk).toBeLessThanOrEqual(1);
        expect(s.phonemeErrorRate).toBeGreaterThanOrEqual(0);
        expect(s.phonemeErrorRate).toBeLessThanOrEqual(0.5);
        expect(s.syllableSkipRate).toBeGreaterThanOrEqual(0);
        expect(s.syllableSkipRate).toBeLessThanOrEqual(0.5);
        expect(s.risk).toBeCloseTo(0.75 * s.syllableSkipRate + 0.25 * s.phonemeErrorRate, 5);
      }
    }
  });

  it('does not tag a fluent reader', () => {
    expect(errorTags(score(meanSquare(0.3), 0))).toEqual([]);
  });

  it('bands the risk score the way the handset does', () => {
    expect(riskLevel(0.7)).toBe('high');
    expect(riskLevel(0.4)).toBe('medium');
    expect(riskLevel(0.2)).toBe('low');
  });
});
